File: maxPathSum.cpp
#include <iostream>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
        
const int INFINT = 0x80000000;
class Solution {
    int maxSum_;
public:
    int calMaxPathSum(TreeNode *root) {
        if (!root) return INFINT;
        int left = INFINT;
        int right = INFINT;
        if (root->left) {
            left = calMaxPathSum(root->left);
        }
        if (root->right) {
            right = calMaxPathSum(root->right);
        }
        int maxsum = std::max(left, right);
        maxsum = maxsum==INFINT ? root->val : max(root->val, root->val+maxsum);
        maxSum_ = max(maxSum_, maxsum);
        if (left != INFINT && right != INFINT)
            maxSum_ = max(maxSum_, root->val+left+right);
        cout << maxSum_ << endl;

        return maxsum;
    }
    int maxPathSum(TreeNode *root) {
        maxSum_ = INFINT;
        calMaxPathSum(root);
        return maxSum_;
    }
};

int main(void) {
    TreeNode *root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    Solution s;
    cout << s.maxPathSum(root) << endl;
    return 0;
}

File: go/tree/maxPathSum.go
package main

import (
	"fmt"
)

// https://leetcode-cn.com/problems/binary-tree-maximum-path-sum/

func maxPathSum(root *TreeNode) int {
	if root == nil {
		return 0
	}

	maxSum := root.Val
	max := func(x, y int) int {
		if x > y {
			return x
		}
		return y
	}
	// calculate the max sum containing node.Val
	var mf func(node *TreeNode) int
	mf = func(node *TreeNode) int {
		if node == nil {
			return 0
		}
		res := node.Val
		left := mf(node.Left)
		right := mf(node.Right)
		res = max(res, res+max(left, right))
		maxSum = max(maxSum, max(res, node.Val+left+right))
		return res
	}
	mf(root)
	return maxSum
}

func main() {
	cases := [][]int{
		{},
		{},
	}

	realCase := cases[0:]
	for i, c := range realCase {
		fmt.Println("## case", i)
		// solve
		fmt.Println(c)
	}
}
